storage/replica_command.go

```diff
@@ -1255,45 +1255,67 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *proto.SplitTrigger) e
 func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDescriptor) (proto.AdminMergeResponse, error) {
 	var reply proto.AdminMergeResponse
 
-	// Lookup subsumed range.
 	if desc.EndKey.Equal(proto.KeyMax) {
-		// Noop.
-		return reply, nil
-	}
-	subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
-	if subsumedRng == nil {
-		return reply, util.Errorf("ranges not collocated; migration of ranges in anticipation of merge not yet implemented")
+		// Merging the final range doesn't make sense.
+		return reply, util.Errorf("cannot merge final range")
 	}
-	subsumedDesc := subsumedRng.Desc()
 
-	// Make sure the range being subsumed follows this one.
-	if !bytes.Equal(desc.EndKey, subsumedDesc.StartKey) {
-		return reply, util.Errorf("Ranges that are not adjacent cannot be merged, %s != %s", desc.EndKey, subsumedDesc.StartKey)
-	}
+	// Lookup subsumed range. This really belongs inside the transaction
+	// for consistency, but it is important (for transaction record placement)
+	// that the first action inside the transaction is the conditional put
+	// to change the left descriptor's end key. We look up the descriptor
+	// here only to get the new end key and then repeat the lookup inside the
+	// transaction.
+	var newEndKey proto.Key
+	{
+		subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
+		if subsumedRng == nil {
+			return reply, util.Errorf("ranges not collocated")
+		}
 
-	// Ensure that both ranges are collocate by intersecting the store ids from
-	// their replicas.
-	if !replicaSetsEqual(subsumedDesc.GetReplicas(), desc.GetReplicas()) {
-		return reply, util.Errorf("The two ranges replicas are not collocate")
+		newEndKey = subsumedRng.Desc().EndKey
+		log.Infof("initiating a merge of %s into %s", subsumedRng, r)
 	}
 
 	// Init updated version of existing range descriptor.
 	updatedDesc := *desc
-	updatedDesc.EndKey = subsumedDesc.EndKey
-
-	log.Infof("initiating a merge of %s into %s", subsumedRng, r)
+	updatedDesc.EndKey = newEndKey
 
 	if err := r.rm.DB().Txn(func(txn *client.Txn) error {
 		// Update the range descriptor for the receiving range.
-		b := &client.Batch{}
-		desc1Key := keys.RangeDescriptorKey(updatedDesc.StartKey)
-		if err := updateRangeDescriptor(b, desc1Key, desc, &updatedDesc); err != nil {
+		{
+			b := &client.Batch{}
+			desc1Key := keys.RangeDescriptorKey(updatedDesc.StartKey)
```

Comment:
desc1Key, can you rename this to match updatedDescKey or something like that to match updatedDesc?